plugin/evm/block_builder.go

```diff
@@ -58,10 +59,11 @@ func (vm *VM) NewBlockBuilder(extraMempool extension.BuilderMempool) *blockBuild
 }
 
 // handleGenerateBlock is called from the VM immediately after BuildBlock.
-func (b *blockBuilder) handleGenerateBlock() {
+func (b *blockBuilder) handleGenerateBlock(currentHeight uint64) {
 	b.buildBlockLock.Lock()
 	defer b.buildBlockLock.Unlock()
-	b.lastBuildTime = time.Now()
+	b.isRetry = b.lastBuiltHeight == currentHeight
```

Comment:
So, `isRetry` will be set to `true` after we have attempted to build a block at the same height 2 times in a row right?

- `waitForEvent` -> `PendingTxs`
- `buildBlock` -> `isRetry = false ; lastBuiltHeight = x`
- `waitForEvent` -> `PendingTxs`
- `buildBlock` -> `isRetry = true ; lastBuiltHeight = x`
- `waitForEvent` -> delay for 500ms

Based on the naming, I would have expected us to be delaying between the first call to `buildBlock` and the second call (rather than after the second call). Specifically:

- `waitForEvent` at `x` -> `PendingTxs`
- `buildBlock` -> `lastBuiltHeight = x ; lastBuiltTime = now()`
- `waitForEvent` at `x` -> delay for 500ms after `lastBuiltTime`

Reply:
updated

Reply:
I think this comment still applies.

I think the only way to delay the first call is if we fetch the currently preferred block in `WaitForEvent` and compare it to the last build block parent hash.

Specifically, I think we should:
- set `lastBuildParentHash` (unconditionally) in `handleGenerateBlock`
- set `lastBuildTime` (unconditionally) in `handleGenerateBlock`
- compare the preferred block hash to the `lastBuildParentHash` in `WaitForEvent`, if they are equal perform the delay based on `lastBuildTime`. Otherwise do not introduce any artificial delay.